---
title: "Introduction to ClustIRR"
author: "S. Kitanovski, K. Wollek"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to ClustIRR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>",
                      fig.height = 5,
                      fig.width = 5,
                      fig.align = "center")
set.seed(987)
```

# Introduction
The T-cell receptor (TCR) is a protein complex found on the surface of T cells 
that is responsible for recognizing antigenic peptides (p) presented in the 
context of the major histocompatibility complex (MHC). Different types of MHCs 
are expressed inside cells and antigen presenting cells (APC), combined with 
antigen peptides to form pMHC complexes, and presented at the cell surface.
If a TCR and pMHC match, complex cell signal interactions can trigger immune
reactions. While MHCs can each bind to a rather broad array of different 
peptides to form pMHCs, TCRs are highly specific to a matching pMHC 
combination. 

B cell receptors (BCRs), on the other hand, are protein complexes 
expressed on B cells that bind directly to antigens. Through immune signalling,
activated B cells can also secrete these receptors as antibodies into the 
blood. Secreted antibodies help the immune system in various ways by binding 
directly to pathogen antigens. In the following we will refer to BCRs 
and TCRs using the general term immune receptors (IRs).

The maturation of B cells and T cells is unique in the human body, in the way 
that partially random recombination of V, D and J genes leads to permanent 
mutations inside the genome of each individual B and T cell.
Owing to VDJ recombination, virtually every B/T cell expresses its own unique 
IR, and hence has a unique specificity. Once a BCR matches to an antigen or a 
TCR to a pMHC structure, an immune reaction can be triggered in form of clonal 
expansion of the matched clonotype, effectively amplifying the number of 
needed pathogen antigen/pMHC specific cells. This diversity is a key element of
the adaptive immune system to deal with constantly evolving pathogens.

Understanding the relationship between the IR sequence and antigens is 
essential for the development of e.g. cancer immunotherapies, vaccines and 
antiviral drugs. By sequencing the VDJ genes of immune cells, expressed IR can 
be reconstructed as amino acid sequences. Certain regions of the receptors
called complementarity determining regions 3 (CDR3) are known to possess high
pMHC binding probability and are therefore mainly focused in specificity 
analysis. Advances in bulk and more currently single cell high-throughput 
sequencing (HT-seq) technologies now enable the analysis of entire repertoires 
of IRs in a single experiment.

Currently available methods for organizing clusters of IR sequences according 
to their likely antigen specificities and quantitative comparison of the 
abundance of the specificity groups between biological conditions or time 
points have crucial drawbacks, or are not available.

This vignette introduces `r Biocpkg("ClustIRR")`, a transparent workflow for 
quantitative exploration of single cell RNA-seq data. 

# ClustIRR algorithm
In short, the algorithm of `r Biocpkg("ClustIRR")` performs clustering of IR 
sequences in a given IR repertoire, to find groups of IRs with similar
specificity. It then employs a quantitative model to compare the abundances 
of the identified IR clusters between the provided input CDR3 sequences and a
reference dataset of CDR3 sequences. In the following each step is explained
using real TCR-seq data as reference.

To run this vignette we need to load a few R-packages:

```{r}
library(ClustIRR)
library(ggplot2)
library(patchwork)
```


# Data

SK: Here we talk about the data that we will analyze. 



# ClustIRR workflow



## 1. Clustering




## 2. ...



Includes a low memory mode to reduce memory usage for large input files. 
(SK: I would mention this later on)

A completely revised version of the 
[**turboGliph**](https://github.com/HetzDra/turboGliph) package focused on the 
clustering part of the **Gliph** algorithm. Input parameters are reduced to
a minimum for user convenience. The first edition of **ClustIRR** implements the 
clustering part of the algorithm. Scoring and plotting parts are planned to be
added in future package versions.

Includes implementations of the 
[**Gliph**](https://github.com/immunoengineer/gliph) and 
[**Gliph2**](http://50.255.35.37:8080) clustering algorithms.
Introduces a new version of the **Gliph** algorithm called `gliph3`,
designed for single-cell sequencing input data.


# Input

Let's use the included **hs_CD8_ref** as reference and **gliph_input_data** 
as sample. This datasets were published with GLIPH.\(^1\)

```{r input_data}
# # load package input data
# data("hs_CD8_ref")
# data("gliph_input_data")
# 
# # detect cores
# cores <- parallel::detectCores()
```

# Clustering

To run **ClustIRR** with default parameters, we provide the columns **CDR3b**,
**TRBV** and **TRBJ** of the data sets (first 3 columns each).

```{r clustering}
# run gliph with default parameters
# gliph_output <- cluster_irr(
# data_sample = gliph_input_data[,1:3],
# data_ref = hs_CD8_ref[,1:3]
# )
```


# Output

**ClustIRR** returns a list of the following elements:

- **clust** _list_: local + global clusters
- **edges** _list_: local + global edges
- **data_sample** _data.frame_: examined data sample
- **version** _integer_: used gliph version
- **ks** _integer vector_: used motif lengths
- **cores** _integer_: number of used CPU cores
- **control** _list_: used auxiliary input parameters

We can for example inspect local clusters:

```{r output}
# inspect output
# head(gliph_output$clust$CDR3b$local$m)

```

- **motif**: found motif
- **f_sample**: motif frequency in the sample set
- **n_sample**: total frequency in the sample set
- **f_ref**: motif frequency in the reference set
- **n_ref**: total motif frequency in the reference set
- **k**: kmer length
- **ove**: motif fold enrichment in the sample set compared to the reference set
- **p_value**: p-value of the Fisher's Exact Test
- **fdr**: random generation probability
- **pass**: selected or not selected motif


# References

**1**: Glanville, Jacob, et al. "Identifying specificity groups in the 
T cell receptor repertoire." Nature 547.7661 (2017): 94.<br>
**2**: Huang, Huang, et al. "Analyzing the Mycobacterium tuberculosis immune 
response by T-cell receptor clustering with GLIPH2 and genome-wide antigen 
screening." Nature Biotechnology 38.10 (2020): 1194-1202.<br>

# Session Info

```{r}
packageVersion("ClustIRR")
```


```{r}
sessionInfo()
```
