---
title: "Introduction to ClustIRR"
author: "S. Kitanovski, K. Wollek"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to ClustIRR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>",
                      fig.height = 5,
                      fig.width = 5,
                      fig.align = "center")
```


```{r}
library(ClustIRR)
library(knitr)
library(visNetwork)
```


# Introduction
To protect the host against genetically diverse and rapidly evolving pathogens, 
adaptive immunity relies on diverse repertoires of immune receptors (IRs), which
includes B- and T-cell receptors (BCRs and TCRs). TCRs recognize peptide:MHC 
complexes, whereas BCRs bind directly to antigens. Antigen recognition may lead 
to immune cell activation. In such a case, the immune cells start to proliferate
rapidly, forming clones of antigen-specific cells capable of mounting effective 
immune response. Hence, decoding the relationship between IRs and antigens is 
essential for developing cancer immunotherapies, vaccines, and antiviral drugs. 

Owing to the advances in high-throughput sequencing technology, it is now 
possible to sequence entire repertoires of IRs. By performing clustering of 
the IR sequences, we can identify clusters of IRs with similar specificity.

This vignette introduces `r Biocpkg("ClustIRR")`, a computational method 
for clustering of immune receptor repertoires.


# ClustIRR algorithm
The algorithm of `r Biocpkg("ClustIRR")` performs clustering of IR sequences 
to find groups of IRs with similar specificity. 

```{r graphic, echo = FALSE, fig.align="left", out.width='90%'}
knitr::include_graphics("../inst/extdata/logo.png")
```



# Input

The main input of ClustIRR are two sets of amino acid sequences of the 
complementarity determining region 3 (CDR3). The CDR3s may come from one T 
cell receptor chain (e.g.\ only CDR3$\alpha$s or only CDR3$\beta$s) or from 
both chains (CDR3$\alpha\beta$). The two sets represent two repertoires:
    
  * `s` : IR repertoire to be analyzed (case sample)
  * `r`: reference IR repertoire (control/reference sample)

**Hint:** `s` and `r` may also consists of CDR3s from $\gamma\delta$ T-cells 
($CDR3\gamma$ and CDR3$\delta$) or B-cells (CDR3H and CDR3L).

Lets look at an example data, which we will use as input:

```{r}
data("CDR3ab")
```

```{r}
# take the first 500 CDR3b sequences from the data -> s
s <- base::data.frame(CDR3b = CDR3ab$CDR3b[1:500])

# take 5000 CDR3b sequences 501:5500 from the data -> r
r <- base::data.frame(CDR3b = CDR3ab$CDR3b[501:5500])
```


```{r}
str(s)
```


```{r}
str(r)
```



# Clustering

`r Biocpkg("ClustIRR")` employs two clustering strategies:

  * **local** clustering: detects enrichment of motifs in the data sample 
    compared to the reference
  * **global** clustering: identifies pairs of CDR3s that have similar
    sequences

The rationale behind these two clustering strategies is the following: two 
identical CDR3 sequences have the same specificity. CDR3 sequences that are
similar (e.g. share the same sequence but differ at only 1 position) likely 
have similar specificity. Global clustering is designed to find such global 
pairs. 

Two CDR3s with significantly different sequences may still recognize the same
peptide if they share a motif in their cores regions (e.g. identical 4-mer ). 
Such "useful" motifs may be enriched in the sample but not in the reference 
repertoire, and the purpose of local clustering is to identify them.



## Local clustering

CDR3 sequences are segmented into overlapping **motifs** ($k$-mers), where $k$ 
is specified by setting the input $ks = 4$. 

Example of segmenting CDR3 sequence into 4-mers:

```{r}
cdr3 <- "CASSTTTGTGELFF"
k <- 4
base::colnames(stringdist::qgrams(cdr3, q = k))
```


$k$-mers found in the core region of the CDR3 loop are more likely to establish 
contact with the antigenic peptide than the $k$-mers found at the flanks of the 
CDR3 sequence. Hence, the user is encouraged to remove few amino acids from the 
flanks of each CDR3 sequence by setting e.g. `control$trim_flank_aa = 3`.

Example of trimming CDR3 flans and segmenting the core of the CDR3 sequence 
into 4-mers:

```{r}
t <- 3
cdr3_trimmed <- base::substr(x = cdr3, start = t+1, stop = nchar(cdr3)-t)
base::colnames(stringdist::qgrams(cdr3_trimmed, q = k))
```
 

A motif is enriched if it satisfies the the following criteria:

  1. `control$local_min_o`: minimum motif frequency in `s`
  2. `control$local_min_ove`: minimum ratio of observed vs. expected (OvE) 
    relative motif frequency, with $OvE=\dfrac{f_s}{n_s}/\dfrac{f_r}{n_r}$
    
      * $f_{s}$ and $f_{r}$: motif frequencies in repertoires `s` and `r`
      * $n_{s}$ and $n_{r}$: total number of motifs in repertoires `s` and `r`
      
  3. `control$local_max_fdr`: maximum false discovery rate (FDR) corrected 
    p-value computed by Fisher's exact test


## Global clustering

For global clustering ClustIRR quantifies the difference between pairs of CDR3
sequences using Hamming distances. Two CDR3 sequences with Hamming distance 
$\leq x$ are considered globally similar, where $x$ is the user-defined 
threshold `control$global_max_dist` (default = 1).

Alternatively, the user can provide a matrix of globally similar CDR3 
sequences computed by a complementary approach (e.g., *tcrdist*) via the 
input parameter `control$global_pairs`.


# Output

`r Biocpkg("ClustIRR")` returns two types of output:

  1. clustering results: tables and lists
  2. graph (igraph object) 


# Case study 1: Detecting enriched motifs

In this study we will insert the motif *LEAR* in the first 20 CDR3 sequences 
of `s`, simulating enrichment of this motif. This motif is not enriched in
repertoire `s` compared to `r`, and ClustIRR should be able to detect *LEAR* 
as enriched:

```{r}
base::substr(x = s$CDR3b[1:20], start = 6, stop = 9) <- "LEAR"
```


... and then we perform clustering with ClustIRR:

```{r}
clustirr_output <- cluster_irr(s = s,  
                               r = r,
                               version = 3,
                               ks = 4,
                               cores = 1,
                               control = list(trim_flank_aa = 3))
```

## Local clustering results
ClustIRR detects *LEAR* as enriched. One additional motifs was detected, 
probably because it contains *LEA* (substring of *LEAR*). *LEAR* was detected
20 times in the `s` and 2 only times in `r` leading to OvE$\approx 100$ 
and FDR$\approx0$.

```{r}
# extract local motifs
local_motifs <- clustirr_output$clust$CDR3b$local$m

# display only passed motifs
knitr::kable(local_motifs[local_motifs$pass == TRUE,], row.names = FALSE)
```

## Global clustering results

Only one pair of globally similar CDR3 sequences was found in our mock data.
The CDR3 sequences *CAS**S**PLEARGYTF* and *CAS**R**PLEARGYTF* differ by one 
amino acid at position 4:

```{r}
# display globally similar pairs
knitr::kable(clustirr_output$clust$CDR3b$global, row.names = FALSE)
```

## Graph

Let's plot the graph:

```{r}
# plot the clust_irr object 
plot_graph(clust_irr = clustirr_output)
```
In this graph, enriched motifs *LEAR* and *LLEA* as well as the global 
connection *CAS-PLEARGYTF* between cells of the sample are visualized. <br>

Nodes can be dragged to improve visibility, if needed. The overall view can be
toggled by clicking and dragging anywhere in the graph outside of nodes. 
Zooming in and out can be done with the mousewheel. <br>

Nodes represent cells. Clicking on a node highlights it's connected edges and 
the CDR3 sequence in the `Select by id` dropdown menu. Hovering over a node 
reveals an information box with the sequence, clone count, and all enriched 
motifs and global connections by chain. Text in the information box can be 
marked for copy&paste. <br>

Edges represent similarity between cells, in form of motifs or global 
connections. Hovering over an edge reveals an information box containing all 
enriched motifs and global connections by chain. Motifs or sequences in the 
information box can be marked for copy&paste. Using the `Select by group` 
dropdown menu, all nodes sharing a motif or globally shared sequence can be 
highlighted. <br>

The legend to the right explains all elements of the graph. In this case,
cells sharing motifs between `CDR3b` sequences appear in blue.
All cells are `Singletons` with a clone count of one and appear as diamonds.
Shared motifs between cells are represented by gray `local edges`, cells 
sharing motifs as well as global sequences are connected by red 
`local & global edges` edges. <br>

To export a graph as for example `.png`-file, the `Export`-button of the Viewer
window can be used after calling the `plot_graph()` function from within the 
console.

# Case study 2: Detecting enriched motifs in clonally expanded IRs

In this case study we assume that the data contains a clone that represents 2% 
(10 T-cells) of the studied repertoire. All T-cells in the expanded clone share 
the CDR3b sequence *CATSRPDGLAQYF*, and bind the same antigenic peptides. 
ClustIRR should detect enrichment of most motifs at the core of *CATSRPDGLAQYF*.

Lets try this out:

```{r}
# create a clone of 10 T-cells 
clone <- base::data.frame(CDR3b = rep("CATSRPDGLAQYF", times = 10))
```

```{r}
# add to the data
s <- base::rbind(s, clone)
```

we perform clustering

```{r}
clustirr_output <- cluster_irr(s = s,  
                               r = r,
                               version = 3,
                               ks = 4,
                               cores = 1,
                               control = list(trim_flank_aa = 3))
```

## Local clustering results
ClustIRR once again reports enrichment of *LEAR*, but also of many additional 
motifs that are part of the core of *CATSRPDGLAQYF*, such as *DGLA*, *PDGL*, 
*RPDG*, *SRPG*, etc.

```{r, fig.align='center'}
# extract local motifs
local_motifs <- clustirr_output$clust$CDR3b$local$m

# display only passed motifs
knitr::kable(local_motifs[local_motifs$pass == TRUE,], row.names = FALSE)
```

## Global clustering results

Once again, ClustIRR finds the same pair of globally similar CDR3 sequences.
However, remember that this time our repertoire `s` contains 10 identical CDR3s 
that belong to the expanded clone. These are also globally similar.
The CDR3 sequences *CAS**S**PLEARGYTF* and *CAS**R**PLEARGYTF* differ by one 
amino acid at position 4:

```{r}
# display globally similar pairs
knitr::kable(clustirr_output$clust$CDR3b$global, row.names = FALSE)
```


## Graph output

Lets plot the clustering results:

```{r}
# plot the clust_irr object 
plot_graph(clust_irr = clustirr_output)
```
This time, the legend shows a dot symbol for clonally `Expanded` cells.
Expanded cells represent clones, i.e. cells that share sequences. <br>

The clonally expanded sequence *CATSRPDGLAQYF* is visible as a big dot.
Dot size represents clone size. Hovering over the big dot in the graph reveals 
also enriched shared motifs between the clones. <br>

To visualize the clone split into single cells (nodes), the `expand_clones`
parameter can be set to `TRUE`:

```{r}
# plot the clust_irr object 
plot_graph(clust_irr = clustirr_output, expand_clones = TRUE)
```

`expand_clones` is set to `FALSE` per default to achieve a more concise 
graph.

# Case study 3: Beta & alpha chains

Example of CDR3$\beta$ and CDR3$\alpha$ sequences in combination. Each cell 
possesses one CDR3$\beta$ and one CDR3$\alpha$ sequence. Shared motifs and 
global similarities are only compared within chains.

We first sample 500 CDR3$\beta$ and CDR3$\alpha$ sequences as input `s`, as well
as 5000 CDR3$\beta$ and CDR3$\alpha$ sequences as reference `r`:

```{r}
data("CDR3ab")

# CDR3a
s_a <- base::data.frame(CDR3a = CDR3ab$CDR3b[4501:5000])
r_a <- base::data.frame(CDR3a = CDR3ab$CDR3b[5001:10000])

# CDR3b
s_b <- base::data.frame(CDR3b = CDR3ab$CDR3b[4501:5000])
r_b <- base::data.frame(CDR3b = CDR3ab$CDR3b[5001:10000])

```

We will insert three clones into each chain as well as two motifs, to 
demonstrate some possible combinations that can occur.

We start with the CDR3$\beta$ chain:

* **Clone 1**: 10 times *CATSRLEFAQYF* 
  * Test global and local edges
* **Clone 2**: 15 times *CATSRLEARATYF* 
  * Test global and local edges + shared motif *LEAR*
* **Clone 2**: 2 times *CATSEEEFAQYF* 
  * Test global edge only
* **Motif 1**: 10 times *LEAR* 
  * Test edge to **Clone 2**
* **Motif 2**: 5 times *REAL* 
  * Will also be CDR3$\alpha$ **Motif 2**, to demonstrate separate clustering

```{r}
# beta clones + motifs
s_b <- base::rbind(s_b, 
                   base::data.frame(CDR3b = rep("CATSRLEFAQYF", times = 10)))
s_b <- base::rbind(s_b, 
                   base::data.frame(CDR3b = rep("CATSRLEARATYF", times = 15)))
s_b <- base::rbind(s_b, 
                   base::data.frame(CDR3b = rep("CATSEEEFAQYF", times = 2)))
base::substr(x = s_b$CDR3b[1:10], start = 4, stop = 7) <- "LEAR"
base::substr(x = s_b$CDR3b[20:25], start = 5, stop = 8) <- "REAL"
```

For the We start with the CDR3$\alpha$ chain we will use the following 
sequences and motifs:

* **Clone 1**: 20 times *CAWNQEEETYEQYF* 
  * Test global and local edges
* **Clone 2**: 5 times *CAWNQEEETYEQYF* 
  * Test global and local edges + shared motif *LEAR*
* **Clone 2**: 2 times *CASSSLYNETTYF* 
  * Test global edge only
* **Motif 1**: 10 times *GPPN* 
  * Test local edges
* **Motif 2**: 5 times *REAL* 
  * Will also be CDR3$\beta$ **Motif 2**, to demonstrate separate clustering

```{r}
# alpha clones + motifs
s_a <- base::rbind(s_a, 
                   base::data.frame(CDR3a = rep("CAWNQEEETYEQYF", times = 20)))
s_a <- base::rbind(s_a, 
                   base::data.frame(CDR3a = rep("CAWNQEEETYEQYF", times = 5)))
s_a <- base::rbind(s_a, 
                   base::data.frame(CDR3a = rep("CASSSLYNETTYF", times = 2)))
base::substr(x = s_a$CDR3a[51:60], start = 6, stop = 9) <- "GPPN"
base::substr(x = s_a$CDR3a[20:25], start = 5, stop = 8) <- "REAL"

```

No we combine both chains:

```{r}
# alpha & beta combined
s_ab <- base::cbind(s_a, s_b)
r_ab <- base::cbind(r_a, r_b)

```

And perform clustering:

```{r}
out3 <- cluster_irr(s = s_ab,
                    r = r_ab,
                    version = 3,
                    ks = 4,
                    cores = 8,
                    control = list(trim_flank_aa = 3))
```
Let's plot the graph:

```{r}
# beta & alpha chain
plot_graph(out3)
```
In this graph, cells that share local motifs and global sequences are connected 
with red edges. Blue nodes signify cells that share local motifs and/or global 
sequences with CDR3$\beta$ sequences of other cells, yellow nodes share with 
CDR3$\alpha$ sequences of other cells and green nodes share both. 

# Case study 4: Gamma & delta chains

A quick example of how CDR3-$\delta&\gamma$ chains can be represented.

We first sample 500 CDR3$\beta$ and CDR3$\alpha$ sequences as input `s`, as well
as 5000 CDR3$\beta$ and CDR3$\alpha$ sequences as reference `r`:

```{r}
data("CDR3ab")

# CDR3g
s_g <- base::data.frame(CDR3g = CDR3ab$CDR3b[4501:5000])
r_g <- base::data.frame(CDR3g = CDR3ab$CDR3b[5001:10000])

# CDR3d
s_d <- base::data.frame(CDR3d = CDR3ab$CDR3a[4501:5000])
r_d <- base::data.frame(CDR3d = CDR3ab$CDR3a[5001:10000])

```

We will again insert three clones into each chain as well as two motifs, to 
demonstrate some possible combinations.

We start with the CDR3$\gamma$ chain:

* **Clone 1**: 4 times *CATSRLEFAQYF* 
  * Test global and local edges
* **Clone 2**: 15 times *CATSRLEARATYF* 
  * Test global and local edges + shared motif *LEAR*
* **Clone 2**: 2 times *CATSEEEFAQYF* 
  * Test global edge only
* **Motif 1**: 10 times *LEAR* 
  * Test edge to **Clone 2**
* **Motif 2**: 5 times *REAL* 
  * Will also be CDR3$\delta$ **Motif 2**, to demonstrate separate clustering

```{r}
# gamma clones + motifs
s_g <- base::rbind(s_g, 
                   base::data.frame(CDR3g = rep("CATSRLEFAQYF", times = 4)))
s_g <- base::rbind(s_g, 
                   base::data.frame(CDR3g = rep("CATSRLEARATYF", times = 15)))
s_g <- base::rbind(s_g, 
                   base::data.frame(CDR3g = rep("CATSEEEFAQYF", times = 2)))
base::substr(x = s_g$CDR3g[1:10], start = 4, stop = 7) <- "LEAR"
base::substr(x = s_g$CDR3g[20:25], start = 5, stop = 8) <- "REAL"

```

Now the CDR3$\delta$ chain:

* **Clone 1**: 4 times *CAWNQEEETYEQYF* 
  * Test global and local edges
* **Clone 2**: 15 times *CAWNQEETTYEQYF* 
  * Test global and local edges + shared motif *LEAR*
* **Clone 2**: 2 times *CASSSLYNETTYF* 
  * Test global edge only
* **Motif 1**: 10 times *GPPN* 
  * Test edge to **Clone 2**
* **Motif 2**: 5 times *REAL* 
  * Will also be CDR3$\gamma$ **Motif 2**, to demonstrate separate clustering

```{r}
# delta clones + motifs
s_d <- base::rbind(s_d, 
                   base::data.frame(CDR3d = rep("CAWNQEEETYEQYF", times = 4)))
s_d <- base::rbind(s_d, 
                   base::data.frame(CDR3d = rep("CAWNQEETTYEQYF", times = 15)))
s_d <- base::rbind(s_d, 
                   base::data.frame(CDR3d = rep("CASSSLYNETTYF", times = 2)))
base::substr(x = s_d$CDR3d[51:60], start = 6, stop = 9) <- "GPPN"
base::substr(x = s_d$CDR3d[20:25], start = 5, stop = 8) <- "REAL"

```

Combine both chains:

```{r}
# gamma & delta combined
s_gd <- base::cbind(s_g, s_d)
r_gd <- base::cbind(r_g, r_d)
```

Perform clustering:

```{r}
out3 <- cluster_irr(s = s_gd,
                    r = r_gd,
                    version = 3,
                    ks = 4,
                    cores = 8,
                    control = list(trim_flank_aa = 3))
```

And plot the graph:

```{r}
plot_graph(out3)
```
CDR3$\gamma$ and CDR3$\delta$ chains are marked as CDR3g and CDR3d in the 
graph. <br>

**Hint:** `s` and `r` may also consists of CDR3s from B-cells (CDR3h and 
CDR3l) by providing *CDR3h* and/or *CDR3l* as input columns.

```{r session_info}
utils::sessionInfo()
```


