---
title: "Introduction to ClustIRR"
author: "S. Kitanovski, K. Wollek"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to ClustIRR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>",
                      fig.height = 5,
                      fig.width = 5,
                      fig.align = "center")
```


# Introduction

This short introduction gives an overview of the biological background of the 
clustering algorithm.

## T cells
The T cell receptor (TCR) is a protein complex found on the surface of T cells 
that is responsible for recognizing antigenic peptides (p) presented in the 
context of the major histocompatibility complex (MHC). Different types of MHCs 
are expressed inside cells and antigen-presenting cells (APC), combined with 
antigen peptides to form pMHC complexes and presented at the cell surface.
If a TCR and a pMHC match, complex cell signal interactions can trigger immune
reactions. While MHCs can each bind to a broad array of different 
peptides to form pMHCs, TCRs are highly specific to a matching pMHC 
combination. 

## B cells
On the other hand, B cell receptors (BCRs) are protein complexes 
expressed on B cells that bind directly to antigens. Through immune signaling,
activated B cells can also secrete these receptors as antibodies into the 
blood. Secreted antibodies help the immune system by binding 
directly to pathogen antigens. In the following, we will refer to BCRs 
and TCRs using the general term immune receptors (IRs).

## V(D)J genes
The maturation of B cells and T cells is unique in the human body in the way 
that partially random recombination of variable (V), diversifying (D), and 
joining (J) gene region with a constant (C) region leads to permanent mutations
inside the genome of each B and T cell. Due to V(D)J recombination, 
virtually every B/T cell expresses a unique IR and has a unique 
specificity. Once a BCR matches an antigen or a TCR to a pMHC structure, an 
immune reaction can be triggered in the form of clonal expansion of the matched 
clonotype, effectively amplifying the number of pathogen antigen/pMHC-specific 
cells. This diversity is a critical element of the adaptive immune system to 
deal with constantly evolving pathogens.

## IR sequencing
Understanding the relationship between the IR sequence and antigens is 
essential for developing, e.g., cancer immunotherapies, vaccines, and 
antiviral drugs. By sequencing the VDJ genes of immune cells, expressed IR can 
be reconstructed as amino acid sequences. Certain regions of the receptors
called complementarity determining regions 3 (CDR3) are known to possess high
pMHC binding probability and are therefore mainly focused on in specificity 
analysis. Advances in bulk and more currently single-cell high-throughput 
sequencing (HT-seq) technologies now enable the analysis of entire repertoires 
of IRs in a single experiment.

## Clustering
Amino acid sequences fold to hypervariable threedimensional structures that 
bind to matching pMHC or antigens. The idea behind clustering is that similar
protein sequences lead to similar protein structures that bind to similar
pMHCs or antigens. Currently, available methods for organizing clusters of IR 
sequences according to their likely antigen specificities and quantitative 
comparison of the abundance of the specificity groups between biological 
conditions or time points have crucial drawbacks or are unavailable.

This vignette introduces `r Biocpkg("ClustIRR")`, a transparent workflow for 
quantitative exploration of single-cell RNA-seq data. 

# ClustIRR algorithm

```{r graphic, echo = FALSE, fig.align="left", out.width='90%'}
knitr::include_graphics("../inst/extdata/logo.png")
```

The algorithm of `r Biocpkg("ClustIRR")` performs clustering of IR 
sequences in a given IR repertoire to find groups of IRs with similar
specificity. It then employs a quantitative model to compare the abundances 
of the identified IR clusters between the provided input CDR3 sequences and a
reference dataset of CDR3 sequences. Each algorithm step is explained in
this vignette using real TCR-seq data as a reference.

`r Biocpkg("ClustIRR")` includes implementations of the Gliph\(^1\) and 
Gliph2\(^2\) clustering algorithms and introduces a third version of the 
Gliph algorithm, designed for single-cell sequencing input data. The version 
can be chosen via the `version` input parameter, default is *version 3*.
See the *version comparison vignette* for details about implementation 
differences. The rest of this vignette will be focused on *version 3* of the 
algorithm.

This first package version focuses on clustering. Input parameters are reduced 
to a minimum for user convenience. Scoring and plotting are planned to be 
added in future versions.

To optimize the execution speed of the algorithm, the input parameter 
`cores` (*default: 1*) can be adjusted to the executing machine (for example 
detectable via `parallel::detectCores()`).

## Input

Two datasets consisting of CDR3 sequences in amino acid code are required 
as input. 

The **sample dataset** has to contain the sequences that should be tested, for 
example, TCR-seq results of a human patient.

The **reference dataset** should be significantly bigger than the sample 
dataset and contain so-called *naive* sequences that are suspected to be 
present in a healthy condition.

Both input datasets should be of type `data.frame` and have to contain one 
column named *CDR3b* or *CDR3a* with amino acid sequences of type character 
(for example, *CASSPVGLYGYTF*). If both columns are supplied within one 
dataset, an additional reference dataset has to be provided containing both 
*CDR3b* and *CDR3a* columns and two runs of the algorithm will 
cluster the sequences separately.

To cluster TCR CDR3δ, TCR CDR3γ, BCR heavy chain, or BCR light chain sequences,
the to be analyzed sequences can be provided in columns named *CDR3b* and 
*CDR3a* to get accepted as input.

The *CD8* dataset that is provided with the package contains 10.000 TCR CDR3β 
chains. It is derived from the T cell receptors of human naive CD8+ T cells and 
was originally published as part of the Gliph2\(^2\) algorithm.

The complete set of sequences is taken as input for clustering.

## Clustering

Two strategies called *local motif clustering* and 
*global CDR3 sequence clustering* are used inside `r Biocpkg("ClustIRR")`.

In short, the goal of local clustering is to find motifs in the data sample 
that are enriched in contrast to the naive reference dataset, while global 
clustering compares CDR3 sequences within the sample dataset for similarity.

In the following subsections motivation and implementation of both 
strategies will get explained.

### Local clustering

CDR3 sequences can be segmented into **motifs** of different lengths. An 
example of a motif of length 4 (also called 4mer) could be the sequence 
*TGTG*, which would be contained by, for example, the sequences *CASSLTGTGYTF*, 
*CASSTTTGTGELFF*, *CSGVTGTGYTF* and many others. A significant abundance of 
similar motifs inside a sample can indicate specificity to similar matching 
pMHC or antigens. To be examined, motif length(s) can be selected via the input
parameter `ks` (*default: c(2,3,4)*).

CDR3 ends at both sides of the IRs are typically found to be non-binding and 
could therefore be excluded from the analysis. The input parameter 
`control$trim_flank_aa` (*default: 0*) can trim all compared sequences 
by *n* amino acids at both sides. For example, an input sequence *CASSLTGTGYTF* 
and `control$trim_flank_aa = 1` would result in *ASSLTGTGYT*, trimmed by *C* at
the start and *F* at the end. If the trim parameter is set too high, a runtime 
warning is produced that shorter sequences were completely trimmed.

A motif is marked as enriched if it passes three filters specified in the input
parameters:

The number of the specific motif found in the sample must be bigger than the
minimum local observation `control$local_min_o` (*default: 1*). It gets 
computed by calculating the frequency of a particular motif in the dataset.
For example, the motif *TGTG* is contained 40 times in the package provided
`CD8` reference dataset.

Furthermore, the Observed vs. Expected (OvE) value must be higher than the 
minimal local OvE threshold `control$local_min_ove` (*default: 2*).
The OvE is calculated with Fisher's exact test using the following formula:

$$ OvE = \frac{f_{sample}}{n_{sample}}/\frac{f_{ref}}{n_{ref}} $$

`f_sample` is motif frequency in the sample, `n_sample` is the count of all 
possible distinct motifs in the data sample of the same length, and `f_ref` as 
well as `n_ref` are the same frequencies in the reference dataset.

Finally, the False Discovery Rate (FDR) of the motif should not exceed the 
maximal local FDR `control$local_max_fdr` (*default: 0.05*).
It is calculated by first computing the probability *p* of finding the 
observed or larger OvE of the motif by change using Fisher's exact test, and 
then adjusting the p-values using the Benjamini & Hochberg (1995) method.

If at least one of these three filters is not passed, the motif is considered 
as not being enriched in the data sample compared to the reference dataset.

### Global clustering

Global clustering compares the similarity of complete CDR3 sequences of the 
same length.

As a pre-processing step before global clustering, the sample dataset is 
reduced to unique CDR3 sequences to prevent self-reference.

For equal-length CDR3 sequences *i* and *j*, the Hamming distance 
*d_ij* can get computed. Sequences with *d_ij* exceeding the global maximal
distance threshold input parameter `control$global_max_dist` (*default: 1*) are 
considered globally similar. For large input files, a low memory mode can be 
turned on via the `control$low_mem` (*default: FALSE*) input parameter, which
reduces memory usage using pairwise comparisons instead of matrix 
computations at the cost of increased computation time.

Alternatively, the user can provide a matrix of globally similar CDR3 
sequences computed by a complementary approach (e.g., *tcrdist*) to save 
computation time. This matrix should contain pairs of CDR3s of the data sample 
set and can be provided via the `control$global_pairs` input parameter.


## Output

`r Biocpkg("ClustIRR")` returns an object of type `clust_irr` that contains 
lists of all found local and global clusters and local edges and 
information about the provided user input. For a detailed list of all values, 
see the *clust_irr manual file*.

### Clustering output

The local motif clustering list contains the motif names and length, a `pass` 
flag, as well as details about computed values inside the algorithm. See the 
*cluster_irr manual file* for a comprehensive list of all listed values.
If, for example, the output was saved as `out` and CDR3b sequences were 
clustered, the motif list can be found under `out$clust$CDR3b$local$m`, and the
local edges list under `out$clust$CDR3b$local$lp`.

The global CDR3 sequence clustering list contains pairs of all sequences marked
as similar, either computer by Hamming Distance or provided either by the user 
as a `global_pairs` input parameter, which at the same time keeps the global
edges between the sequences. If, for example, the output was saved as `out` and
CDR3b sequences were clustered, the CDR3 sequence pairs can be found under 
`out$clust$CDR3b$global`.

### Graph output

A function to plot a graph directly is planned to be added to future package 
versions.

The `get_edges()` function can be used to extract a list of local and global
connections between the CDR3 sequences of the sample dataset. 

# Case study

A short case study is conducted to demonstrate the `r Biocpkg("ClustIRR") 
workflow. The newest *version 3* will be used. For a comparison of the 
different algorithm versions, see the *version comparison vignette*.

We start by loading the needed packages. *knitr* will be used to display the 
results in tabular form.

```{r case_study}
# load packages
library(ClustIRR)
library(knitr)
```

## Input data
A randomly sampled part of the reference dataset will be used as a sample 
dataset. To test the algorithm against ground truth, the sample dataset will be
artificially enriched with the motif *RQWW* and with the two through copying 
simulated clonal expanded sequences *CATSRAAKPDGLAALETQYF* and 
*CATSRAAKPDGLAALSTQYF*.

The package provided `CD8` dataset is loaded as an example reference dataset. 
This set of 10.000 β-chains is derived from the T cell receptors of human naive
CD8+ T cells was published as part of the Gliph 2\(^2\) algorithm.


```{r input_data}
# load package input data
data("CDR3ab")
CDR3ab$CDR3a <- NULL

# set the random seed for sampling reproducibility
set.seed(987)

# sample 500 sequences from the reference dataset as sample dataset
data_sample <- data.frame(CDR3b = CDR3ab[sample(x = 1:nrow(CDR3ab), 
                                                size = 500, 
                                                replace = FALSE),])

# artificially enrich motif 'RQWW' inside sample dataset
substr(x = data_sample$CDR3b[1:20], start = 6, stop = 9) <- "RQWW"

# add the artificial clonal expansion of two sequences to the sample dataset
clones <- data.frame(CDR3b = rep(x = c("CATSRAAKPDGLAALETQYF",
                                       "CATSRAAKPDGLAALSTQYF"),
                                 times = 15))
data_sample <- rbind(data_sample, clones)

```

## Clustering

Only the sample and reference
dataset must be provided as input to perform clustering with the default 
values. For demonstration purposes, we will only look for motifs of length 
`ks` = 4 and trim the flank amino acids by three on each side.

```{r clustering}
# perform clustering
clustirr_output <- cluster_irr(s = data_sample,  
                               r = CDR3ab,
                               version = 3,
                               ks = 4,
                               cores = 1,
                               control = list(
                                 B = 1000,
                                 global_max_dist = 1,
                                 local_max_fdr = 0.05,
                                 local_min_ove = 2,
                                 local_min_o = 3,
                                 trim_flank_aa = 3,
                                 low_mem = FALSE,
                                 global_pairs = NULL))

```

A warning tells us that CDR3 sequences of the reference dataset shorter than 
six amino acids were trimmed completely before local clustering was performed.
This happens because we set the `trim_flank_aa` parameter to 3, so at each 
side of all CDR3 sequences, three aa got cut off, effectively trimming 
sequences $\leq 6$ aa completely. This should be fine as long as our sample 
dataset is not cut.

## Output

### Local motif clustering output

Inspecting the local motif clustering shows that, in total, 20 motifs were 
marked as enriched. The 20 times inserted motif *RQWW* gets found 
twenty times (to be seen in the column `f_sample`). Five other motifs are 
combinations of the first three (*RQW* <-> *QRQW*, *YRQW*) or last three 
(*QWW* <-> *QWWA*, *QWWG*, *QWWS*) amino acids of the inserted motif with the 
respective sequences it was inserted in. 14 motifs of the clonally expanded 
sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF* are found. Some are
found in both sequences (for example, *AAKP*, found 30 times), and others only 
in one of the two (for example, *AALE*, found 15 times). Notice that 
motifs can also be present in the reference dataset as parts of other sequences 
(`f_ref` > 0). For example, the enriched motif *GLAA* was 
contained in the sample before the artificial enhancement, leading to a count 
of 31. 

If the random seed value changed, slightly different motif counts 
could appear through interaction with sampled motifs of the reference set. 
All inserted motifs and shared motifs between the 
clonally expanded sequences are detected as enriched.

```{r local_motifs_output}
# extract local motifs
local_motifs <- clustirr_output$clust$CDR3b$local$m

# display only passed motifs
kable(local_motifs[local_motifs$pass == TRUE,])

```

### Global CDR3 sequence clustering output

Inspecting the global motif clustering reveals two found sequence pairs of 
CDR3 sequences.

The first pair, *CASSYVLRAADGYTF* and *CASSYVLRAADGDTF*, was present in the 
reference dataset and were randomly sampled into the example dataset.

The second pair *CATSRAAKPDGLAALSTQYF* and *CATSRAAKPDGLAALETQYF* represents 
the artificially enhanced clonal expansion.

Again, if the sampling was repeated with a different seed, other randomly 
similar pairs of the reference dataset could be sampled. The two added 
sequences are guaranteed to be found.


```{r global_motifs_output}
# global motifs
kable(clustirr_output$clust$CDR3b$global)
```
### Graph output

#### Extract edges

To extract the edges for a visualization, the `get_edges()` function can be 
called on our output file of class `clust_irr`.


```{r local_edges_output}
# extract edges
edges <- get_edges(clustirr_output)

```

#### Graph output

As mentioned earlier, an integrated plotting function is planned for future
`r Biocpkg("ClustIRR")` versions.

# References

**1**: Glanville, Jacob, et al. "Identifying specificity groups in the 
T cell receptor repertoire." Nature 547.7661 (2017): 94.<br>
**2**: Huang, Huang, et al. "Analyzing the Mycobacterium tuberculosis immune 
response by T-cell receptor clustering with GLIPH2 and genome-wide antigen 
screening." Nature Biotechnology 38.10 (2020): 1194-1202.<br>

# Session Info

```{r package_version}
packageVersion("ClustIRR")
```


```{r session_info}
sessionInfo()
```
