---
title: "Introduction to ClustIRR"
author: "S. Kitanovski, K. Wollek"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Introduction to ClustIRR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>",
                      fig.height = 5,
                      fig.width = 5,
                      fig.align = "center")
set.seed(987)
```

# Introduction
The T-cell receptor (TCR) is a protein complex found on the surface of T cells 
that is responsible for recognizing antigenic peptides (p) presented in the 
context of the major histocompatibility complex (MHC). Different types of MHCs 
are expressed inside cells and antigen presenting cells (APC), combined with 
antigen peptides to form pMHC complexes, and presented at the cell surface.
If a TCR and pMHC match, complex cell signal interactions can trigger immune
reactions. While MHCs can each bind to a rather broad array of different 
peptides to form pMHCs, TCRs are highly specific to a matching pMHC 
combination. 

B cell receptors (BCRs), on the other hand, are protein complexes 
expressed on B cells that bind directly to antigens. Through immune signalling,
activated B cells can also secrete these receptors as antibodies into the 
blood. Secreted antibodies help the immune system in various ways by binding 
directly to pathogen antigens. In the following we will refer to BCRs 
and TCRs using the general term immune receptors (IRs).

The maturation of B cells and T cells is unique in the human body, in the way 
that partially random recombination of V, D and J genes leads to permanent 
mutations inside the genome of each individual B and T cell.
Owing to VDJ recombination, virtually every B/T cell expresses its own unique 
IR, and hence has a unique specificity. Once a BCR matches to an antigen or a 
TCR to a pMHC structure, an immune reaction can be triggered in form of clonal 
expansion of the matched clonotype, effectively amplifying the number of 
needed pathogen antigen/pMHC specific cells. This diversity is a key element of
the adaptive immune system to deal with constantly evolving pathogens.

Understanding the relationship between the IR sequence and antigens is 
essential for the development of e.g. cancer immunotherapies, vaccines and 
antiviral drugs. By sequencing the VDJ genes of immune cells, expressed IR can 
be reconstructed as amino acid sequences. Certain regions of the receptors
called complementarity determining regions 3 (CDR3) are known to possess high
pMHC binding probability and are therefore mainly focused in specificity 
analysis. Advances in bulk and more currently single cell high-throughput 
sequencing (HT-seq) technologies now enable the analysis of entire repertoires 
of IRs in a single experiment.

Currently available methods for organizing clusters of IR sequences according 
to their likely antigen specificities and quantitative comparison of the 
abundance of the specificity groups between biological conditions or time 
points have crucial drawbacks, or are not available.

This vignette introduces `r Biocpkg("ClustIRR")`, a transparent workflow for 
quantitative exploration of single cell RNA-seq data. 

# ClustIRR algorithm
In short, the algorithm of `r Biocpkg("ClustIRR")` performs clustering of IR 
sequences in a given IR repertoire, to find groups of IRs with similar
specificity. It then employs a quantitative model to compare the abundances 
of the identified IR clusters between the provided input CDR3 sequences and a
reference dataset of CDR3 sequences. Each step of the algorithm is explained
using real TCR-seq data as reference.

To run this vignette the following R-packages are used:

```{r}
library(ClustIRR)
library(ggplot2)
library(patchwork)
```


# Data

Two datasets consisting of CDR3 sequences in amino acid code are needed 
as input:

* one **sample dataset** that should be tested
* one naive **reference dataset** to test the sample dataset against

As example reference dataset, the package provided \code{CD8} dataset is loaded. 
This set of 10000 \eqn{\beta}-chains is derived from the T cell receptors of 
human naive CD8+ T cells was published as part of the GLIPH 2\(^2\) algorithm.

As sample dataset a randomly sampled part of the reference dataset will be used.
To test the algorithm against ground truth, the sample dataset will be
artificially enriched with the motif *RQWW* and clonal expanded with the two
sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*.


```{r input_data}
# load package input data
data("CD8")

# sample 500 sequences from the reference dataset as sample dataset
data_sample <- data.frame(CDR3b = CD8[sample(x = 1:nrow(CD8), 
                                             size = 500, 
                                             replace = FALSE),])

# artificially enrich motif 'RQWW' inside sample dataset
substr(x = data_sample$CDR3b[1:20], start = 6, stop = 9) <- "RQWW"

# add artificial clonal expansion of two sequences to sample dataset
clones <- data.frame(CDR3b = rep(x = c("CATSRAAKPDGLAALETQYF",
                                       "CATSRAAKPDGLAALSTQYF"),
                                 times = 15))
data_sample <- rbind(data_sample, clones)

```

# ClustIRR workflow

`r Biocpkg("ClustIRR")` includes implementations of the 
[**Gliph**](https://github.com/immunoengineer/gliph) and 
[**Gliph2**](http://50.255.35.37:8080) clustering algorithms and
introduces a new version of the **Gliph** algorithm called `gliph3`,
designed for single-cell sequencing input data.

The first version of `r Biocpkg("ClustIRR")` is focused on the clustering part 
of the **Gliph** algorithms. Input parameters are reduced to a minimum for user 
convenience. A low memory mode to reduce memory usage for large input files
is included. Scoring and plotting parts are planned to be added in future 
versions.

## Clustering

Two strategies are used to cluster the CDR3 sequences:

* Local motif clustering
* Global CDR3 sequence clustering

### Local clustering

CDR3 sequences can be segmented into **motifs** of a certain length. Significant 
abundance of similar motifs inside a sample can indicate specificity to 
matching pMHC or antigens. To cluster the sequences the following steps are 
performed:

1. CDR3 pre-processing
  * Get set of non-redundant CDR3s *(only in version = 1 or 2)*
  * Trim CDR3 ends, that are typically non-binding
  
2. Motif processing
  * count motif frequencies in sample *f_sample* and reference *f_ref*
  * count total number of motifs in sample *n_sample* and reference *n_ref*
  * compute ratio of observed vs. expected motif counts using the following 
  formula:
    * *OvE* = (*f_sample* / *n_sample* ) / ( *f_ref* / *n_ref* )
  * In version 2 and 3, probability *p_i* of finding the observed or a larger 
  *OvE* for motif *i* given that the null hypothesis is true is computed with 
  the Fisher's exact test
  * In version 1, *p_i* is computed by bootstrapping
  * classify motif *i* as passed, if the motif passes all filters specified in 
  the user-provided control list

### Global clustering

`r Biocpkg("ClustIRR")` provides two possible approaches for global clustering:

1. For equal-length CDR3 sequences *i* and *j* compute their Hamming distance *d_ij*. Sequences with *d_ij* *global_max_dist*
(user-defined input) are considered globally similar

2. Alternatively, the user can provide a matrix of globally similar CDR3 sequences computed by a complementary approach (e.g. *tcrdist*)

## Case study

```{r clustering}
# detect cores of the executing computer 
cores <- parallel::detectCores()

out <- cluster_irr(data_sample = data_sample,
                   data_ref = CD8,
                   ks = 4,
                   cores = cores,
                   version = 3)


```

## Output
**ClustIRR** returns a list of the following elements:

- **clust** _list_: local + global clusters
- **edges** _list_: local + global edges
- **data_sample** _data.frame_: examined data sample
- **version** _integer_: used gliph version
- **ks** _integer vector_: used motif lengths
- **cores** _integer_: number of used CPU cores
- **control** _list_: used auxiliary input parameters

We can for example inspect local clusters:

```{r output}
# inspect output
# head(gliph_output$clust$CDR3b$local$m)

```

- **motif**: found motif
- **f_sample**: motif frequency in the sample set
- **n_sample**: total frequency in the sample set
- **f_ref**: motif frequency in the reference set
- **n_ref**: total motif frequency in the reference set
- **k**: kmer length
- **ove**: motif fold enrichment in the sample set compared to the reference set
- **p_value**: p-value of the Fisher's Exact Test
- **fdr**: random generation probability
- **pass**: selected or not selected motif

# References

**1**: Glanville, Jacob, et al. "Identifying specificity groups in the 
T cell receptor repertoire." Nature 547.7661 (2017): 94.<br>
**2**: Huang, Huang, et al. "Analyzing the Mycobacterium tuberculosis immune 
response by T-cell receptor clustering with GLIPH2 and genome-wide antigen 
screening." Nature Biotechnology 38.10 (2020): 1194-1202.<br>

# Session Info

```{r}
packageVersion("ClustIRR")
```


```{r}
sessionInfo()
```
