---
title: "Version comparison"
author: "S. Kitanovski, K. Wollek"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_depth: 4
vignette: >
  %\VignetteIndexEntry{Version comparison}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      comment = "#>",
                      fig.height = 5,
                      fig.width = 5,
                      fig.align = "center")
```

# ClustIRR algorithm version differences

`r Biocpkg("ClustIRR")` offers implementations of the Gliph\(^1\) and 
Gliph2\(^2\) clustering algorithms and introduces a new, third version of the 
Gliph algorithm, designed for single-cell sequencing input data. The version 
can be chosen via the `version` input parameter (*default: version 3*).

This vignette will focus on the differences between the versions. For an 
overall introduction into what `r Biocpkg("ClustIRR")` does and the biological 
and theoretical background behind the algorithm see the 
*user manual vignette*.

## Input

As a pre-processing step before local clustering, *version 1* and *version 2* 
of the algorithm reduce both the sample dataset and the reference dataset to 
unique sequences to reduce the examined clonally expanded sequences to only 
clonotypes. Neither the Gliph\(^1\) nor the Gliph2\(^2\) paper contains 
the reasoning behind this. Information about clonal 
expansion was only considered in the algorithm's later scoring stage. One 
possible explanation could be that clonal expansion counts 
were suspected to be biased, induced by the predominantly used 
bulk sequencing techniques at the time. As a result, this leads to a 
systematic negative bias in the local clustering stage against the motifs in 
clonally expanded sequences, which could arguably be the most important ones 
for specificity analysis.

*Version 3* of the algorithm takes the complete set of sequences as input for 
local clustering, taking into account that newer single-cell high-throughput 
sequencing techniques are less biased regarding clonal expansion counts.

Therefore, *version 3* should be used when single-cell sequencing data is
available for analysis, while *version 2* and *version 1* of the algorithm 
should be used for analyzing bulk sequencing data.

Be aware that *version 1* potentially produces false positives and should only 
be used to reproduce results from papers that used Gliph\(^1\). The case study 
below will give an example of falsely detected motifs and explain why 
these are found.

## Clustering

In the current `r Biocpkg("ClustIRR")` implementation, the three versions of 
the algorithm differ only in the *local motif clustering* part. The algorithm
for *global CDR3 sequence clustering* is identical for all versions.

### Local clustering

In *version 2 & 3*, the OvE is calculated with Fisher's exact test:

$$ OvE = \frac{f_{sample}}{n_{sample}}/\frac{f_{ref}}{n_{ref}} $$

In *version 1*, the OvE is computed by bootstrapping, i.e., repeated sampling 
of the reference dataset with a sample size equal to the sample dataset. 
Sampling depth can be configured for higher precision by changing the input 
parameter `control$B` (*default: 1000*) to higher values, which only affects 
*version 1*. Using too low precision can produce false positives.

The False Discovery Rate (FDR) is calculated by first computing the 
probability *p* of finding the observed or larger OvE of the motif by change 
using bootstrapping in *version 1* or Fisher's exact test in *version 2 & 3*, 
and then adjusting the p-values using the Benjamini & Hochberg (1995) method.

### Global clustering

As a pre-processing step before global clustering, all versions reduce the 
sample dataset to unique CDR3 sequences to prevent self-reference.

## Output

All `r Biocpkg("ClustIRR")` versions return an object of type `clust_irr`.

### Clustering output

Because *version 1* uses bootstrapping and *version 2 & 3* Fisher's exact test,
a few output values are different between the versions:

*Version 2 & 3* report additional confidence intervals for the OvE value.

*Version 1* reports the motif counts found in the sample with additional 
minimum and maximum found motifs.

See the *cluster_irr manual file* for details on all output values.

### Graph output

Found edges and nodes can get extracted via the `get_edges` function. 
*Version 3* also detects motifs shared between clonally expanded sequences, 
which should be kept in mind before building graphs.

# Case study

To demonstrate the differences, we will conduct the same case 
study as integrated into the *user manual vignette* with all three versions.

We start by loading the needed packages. *knitr*, *ggplot2* and *ggrepel* are
used to display the results.

```{r case_study}
# load packages
library(ClustIRR)
library(knitr)
library(ggplot2)
library(ggrepel)
```

## Input data
Like in the *user manual vignette*, a randomly sampled part of the reference 
dataset will be used as a sample dataset. To test the algorithm against the 
ground truth, the sample dataset will be artificially enriched with the motif 
*RQWW*, which will get inserted into 20 sequences, and the two sequences 
*CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*, which will be added 15 times
each to simulate clonal expansion.

The package provided `CD8` dataset is loaded as an example reference dataset.

```{r input_data}
# load package input data
data("CD8")

# set the random seed for sampling reproducibility
set.seed(987)

# sample 500 sequences from the reference dataset as sample dataset
data_sample <- data.frame(CDR3b = CD8[sample(x = 1:nrow(CD8), 
                                             size = 500, 
                                             replace = FALSE),])

# artificially enrich motif 'RQWW' inside sample dataset
substr(x = data_sample$CDR3b[1:20], start = 6, stop = 9) <- "RQWW"

# add the artificial clonal expansion of two sequences to the sample dataset
clones <- data.frame(CDR3b = rep(x = c("CATSRAAKPDGLAALETQYF",
                                       "CATSRAAKPDGLAALSTQYF"),
                                 times = 15))
data_sample <- rbind(data_sample, clones)

```

## Clustering

We will again only look for motifs of length `ks` = 4. This time, outputs for 
all three versions of the algorithm are computed, i.e., *version 3*, 
*version 2*, and *version 1*. All input parameters between the versions are set
to identical values.

```{r clustering}
# ClustIRR version 3
clustirr_output_v3 <- cluster_irr(data_sample = data_sample,  
                                  data_ref = CD8,
                                  version = 3,
                                  ks = 4,
                                  cores = 1,
                                  control = list(
                                    B = 1000,
                                    global_max_dist = 1,
                                    local_max_fdr = 0.05,
                                    local_min_ove = 2,
                                    local_min_o = 3,
                                    trim_flank_aa = 3,
                                    low_mem = FALSE,
                                    global_pairs = NULL))
# ClustIRR version 2
clustirr_output_v2 <- cluster_irr(data_sample = data_sample,  
                                  data_ref = CD8,
                                  version = 2,
                                  ks = 4,
                                  cores = 1,
                                  control = list(
                                    B = 1000,
                                    global_max_dist = 1,
                                    local_max_fdr = 0.05,
                                    local_min_ove = 2,
                                    local_min_o = 3,
                                    trim_flank_aa = 3,
                                    low_mem = FALSE,
                                    global_pairs = NULL))
# ClustIRR version 1
clustirr_output_v1 <- cluster_irr(data_sample = data_sample,  
                                  data_ref = CD8,
                                  version = 1,
                                  ks = 4,
                                  cores = 1,
                                  control = list(
                                    B = 1000,
                                    global_max_dist = 1,
                                    local_max_fdr = 0.05,
                                    local_min_ove = 2,
                                    local_min_o = 3,
                                    trim_flank_aa = 3,
                                    low_mem = FALSE,
                                    global_pairs = NULL))


```
Warnings tell us that CDR3 sequences of the reference datasets shorter than 
six amino acids were trimmed completely before local clustering was performed.
This happens because we set the `trim_flank_aa` parameter to 3, so at each 
side of all CDR3 sequences, three aa got cut off, effectively trimming 
sequences $\leq 6$ aa completely. This should be fine as long as our sample 
datasets are not cut.

## Output

In this section, we will compare the output of the different versions in 
detail.

While the three `r Biocpkg("ClustIRR")` versions of the algorithm differ only 
in the *local motif clustering* part, the *global CDR3 sequence clustering* is
identical for all versions and produces the same output.

### Local motif clustering output

Each version will be inspected separately and compared to the other versions 
for local clustering. We will start with the newest version of the algorithm.

#### Version 3

Inspecting the local motif clustering of *version 3* shows that, in total, 20 
motifs were marked as enriched. 

Seven motifs relating to the 20 times inserted motif get found. The inserted 
motif *RQWW* gets found 20 times (`f_sample` column). Five other motifs are 
combinations of the first three (*RQW* <-> *QRQW*, *YRQW*) or last three 
(*QWW* <-> *QWWA*, *QWWG*, *QWWS*) amino acids of the inserted motif with the 
respective sequences it was inserted in. The motif *SYRQ* comprises the 
first two digits *RQ* of the inserted motif and *SY* of the four cdr3 sequences
it was inserted into.

13 motifs related to the clonally expanded sequences *CATSRAAKPDGLAALETQYF* 
and *CATSRAAKPDGLAALSTQYF* are found. Some are found in both sequences 
(for example, *AAKP*, found 30 times), and some only in one of the two (for
example, *AALE*, found 15 times). Notice that several of the newly found motifs 
were also present in the reference dataset as parts of other sequences 
(`f_ref` > 0), and in this case, by chance, one of them (*GLAA*) was contained 
in the sample before the artificial enhancement, leading to a count of 31. 

If the random seed value changed, slightly different motif counts 
could appear through interaction with sampled motifs of the reference set. 
In any case, the inserted motifs and the motifs found in the clonally 
expanded sequences are guaranteed to be detected by the algorithm *version 3*.


```{r local_motifs_output_version_3}
# extract local motifs version 3
local_motifs_v3 <- clustirr_output_v3$clust$CDR3b$local$m

# display passed motifs
kable(local_motifs_v3[local_motifs_v3$pass == TRUE,])

```

#### Version 2

Inspecting the local motif clustering of *version 2* shows that, in total, 6 
motifs were marked as enriched. The 20 times inserted motif *RQWW* gets found 
20 times(`f_sample` column). The other five motifs are all combinations of the 
first three (*RQW* <-> *QRQW*, *YRQW*) or the last three (*QWW* <-> *QWWA*, 
*QWWG*, *QWWS*) amino acids of the inserted motif with the respective sequences
it was inserted in.

Because of the redundancy reduction performed in *version 2*, no motifs of the 
2 clonal expanded sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF* 
are found, which both were inserted 15 times into the sample data.


```{r local_motifs_output_version_2}
# extract local motifs version 2
local_motifs_v2 <- clustirr_output_v2$clust$CDR3b$local$m

# display passed motifs
kable(local_motifs_v2[local_motifs_v2$pass == TRUE,])
```

#### Version 1

Inspecting the local motif clustering of *version 1* shows that, in total, 12 
motifs were marked as enriched. The 20 times inserted motif *RQWW* gets found 
twenty times (`f_sample` column). Five other motifs are combinations of the 
first three amino acids (*RQW* <-> *QRQW*, *YRQW*) or the last three amino 
acids (*QWW* <-> *QWWA*, *QWWG*, *QWWS*) amino acids of the inserted motif with
the respective sequences it was inserted in. The motif *SYRQ* comprises the 
first two digits *RQ* of the inserted motif and *SY* of the four cdr3 sequences
it was inserted into.

Because of the redundancy reduction performed in *version 1*, no motifs of the 
2 clonal expanded sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF* 
are found, which both were inserted 15 times into the sample data.

*Version 1* uses bootstrapping to compute enrichment (*version 3* and 
*version 2* use Fisher's exact test, see man-files for details). 
This can induce noise, as repeated samples are drawn randomly, and also limits
the p-value precision.

The remaining 5 motifs *SNRA*, *SLVS*, *LVSG*, *GGSY* and *GFNT* are
found in this case, but if the random seed is changed, other motifs appear
as enriched.

```{r local_motifs_output_version_1}
# extract local motifs version 1
local_motifs_v1 <- clustirr_output_v1$clust$CDR3b$local$m

# display passed motifs
kable(local_motifs_v1[local_motifs_v1$pass == TRUE,])
```

#### Version 3 vs. Version 2

For a direct comparison between the versions, we will plot the motifs using the
p-values of the different versions.
To enhance the visibility of the data in the graph, the observed values were 
subjected to a negative log10 transformation.

We start by comparing *version 3* and *version 2*.

```{r local_motifs_v3_vs_v2, fig.height=7, fig.width=11}
# add version suffixes to all columns except the motif column
names(local_motifs_v1)[names(local_motifs_v1) != "motif"] <- 
  paste(names(local_motifs_v1)[names(local_motifs_v1) != "motif"], 
        "v1", sep="_")
names(local_motifs_v2)[names(local_motifs_v2) != "motif"] <- 
  paste(names(local_motifs_v2)[names(local_motifs_v2) != "motif"], 
        "v2", sep="_")
names(local_motifs_v3)[names(local_motifs_v3) != "motif"] <- 
  paste(names(local_motifs_v3)[names(local_motifs_v3) != "motif"], 
        "v3", sep="_")

# combine the output of all versions for easier filtering
local_motifs <- Reduce(function (...) { 
  merge(..., by = "motif", all = TRUE) },
  list(local_motifs_v1, local_motifs_v2, local_motifs_v3)) 

# select all motifs that are passed in at least one version of v2 and v3
local_motifs_v3_v2 <- local_motifs[(local_motifs$pass_v2 == TRUE |
                                      local_motifs$pass_v3 == TRUE),]

# display with ggplot
ggplot(local_motifs_v3_v2, aes(-log10(p_value_v3),
                               -log10(p_value_v2),
                               label = motif,
                               color = pass_v2)) +
  geom_point(size=3) +
  geom_label_repel(nudge_y = 2, force = 100, max.overlaps = Inf, 
                   show.legend = FALSE, size = 7) +
  scale_colour_manual(name="Found motifs",
                      label=c("v3 only (14)", 
                              "both v2 & v3 (6)"),
                      values= c("darkgray",
                                "black")) +
  labs(x="-log10(p-value v3)", y = "-log10(p-value v2)") +
  theme(text = element_text(size = 23)) +
  lims(x = c(0,50), y=c(0,30))

```

The inserted motif *RQWW* gets detected with a high p-value by both versions, 
as well as the related motifs *QWWA*, *QWWG*, *QWWS*, *QRQW*, and *YRQW* 
(marked black in the graph). The motif *SYRQ* is only detected as enriched by 
*version 3*. This is caused by a slightly higher false discovery rate value 
calculated by *version 2* (*0.05268845*) than the cutoff of *0.05*. Reducing 
the sequences to a non-redundant set also affects the number of motifs present
in the data sample and reference sample, leading to *version 2* p-value 
correction values being different than in *version 3*, where this reduction is
not part of the algorithm.

Thirteen shared motifs between clonally expanded sequences are only marked as 
enriched by *version 3* (marked dark gray in the graph). *Version 2* reduces 
non-redundant sequences before looking for motifs and therefore does not detect
motifs between clonally expanded sequences as enriched.

The motifs *AAKP*, *AALE*, *AALS*, *AKPD*, *ALET*, *ALST*, *DGLA*, *GLAA*, 
*KPDG* ,*LAAL*, *PDGL*, *RAAK* and *SRAA* are all part of the two clonally 
expanded sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*.

In total, *version 2* finds 6 motifs, while *version 3* finds 20 motifs: the 6 
motifs also found by *version 2*, as well as 1 additional motif related to 
motif insertion of *RQWW*, and 13 clonal expansion-related motifs.

#### Version 3 vs. Version 1

Next, we will compare *version 3* with *version 1*.

```{r local_motifs_v3_v_v1, fig.height=7, fig.width=11}
# select all motifs that are passed in at least one version of v1 and v3
local_motifs_v3_v1 <- local_motifs[(local_motifs$pass_v1 == TRUE |
                                      local_motifs$pass_v3 == TRUE),]

# calculate pass_v1_v3 parameter for visualization
local_motifs_v3_v1$pass_v1_v3 <- "none"
local_motifs_v3_v1$pass_v1_v3 <-
  base::ifelse(test = local_motifs_v3_v1$pass_v3,
               yes = "v3",
               no = local_motifs_v3_v1$pass_v1_v3)
local_motifs_v3_v1$pass_v1_v3 <-
  base::ifelse(test = local_motifs_v3_v1$pass_v1,
               yes = "v1",
               no = local_motifs_v3_v1$pass_v1_v3)
local_motifs_v3_v1$pass_v1_v3 <-
  base::ifelse(test = local_motifs_v3_v1$pass_v3 &
                 local_motifs_v3_v1$pass_v1,
               yes = "both",
               no = local_motifs_v3_v1$pass_v1_v3)

# display with ggplot
ggplot(local_motifs_v3_v1, aes(-log10(p_value_v3),
                               -log10(p_value_v1),
                               label = motif,
                               color = pass_v1_v3)) +
  geom_point(size=3) +
  geom_label_repel(nudge_x = 3, force = 100, max.overlaps = Inf, 
                   show.legend = FALSE, size = 7) +
  scale_colour_manual(name="Found motifs",
                      label=c("both v1 & v3 (7)", 
                              "v1 only (5)", 
                              "v3 only (13)"),
                      values = c("black", 
                                 "red", 
                                 "darkgray")) +
  labs(x="-log10(p-value v3)", y = "-log10(p-value v1)") +
  theme(text = element_text(size = 23)) +
  lims(x = c(0,50), y=c(0,30))

```

The inserted motif *RQWW* gets detected with a high p-value by both versions, 
as well as the related motifs *QWWA*, *QWWG*, *QWWS*, *QRQW* and *YRQW* 
and the motif *SYRQ* (marked black in the graph). This is in case of 
*version 1* caused by underestimation of the number of found motifs in the 
reference dataset (`mean_f_ref`), which is calculated as the mean of repeated 
bootstrapping, leading in effect to a "happy accident" in which *SYRQ* is found 
by chance.

13 shared motifs between clonally expanded sequences are only marked as 
enriched by *version 3* (marked dark gray in the graph). *Version 1* reduces 
like *Version 2* non-redundant sequences before looking for motifs and 
therefore does not detect motifs between clonally expanded sequences as 
enriched.

The motifs *AAKP*, *AALE*, *AALS*, *AKPD*, *ALET*, *ALST*, *DGLA*, *GLAA*, 
*KPDG* ,*LAAL*, *PDGL*, *RAAK* and *SRAA* are all part of the two clonally 
expanded sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*.

Five additional motifs are only found by *version 1* (marked red in the graph).
The found 4mers *GFNT*, *GGSY*, *LVSG*, *SLVS*, and *SNRA* are not related
to the artificial enrichment of motif *RQWW* but also can't be found in the two
clonally expanded sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*. 
The rest of the sample dataset is a part of the naive reference dataset it was
created from, which does not contain enriched motifs. Therefore, these motifs 
represent false positives caused by the p-value precision of *version 1* being
too low.

In total, *version 1* finds 12 motifs, of which five motifs are false positives 
and 7 are motifs related to the inserted motif *RQWW*.
*Version 3* finds 20 motifs, containing all seven motifs found by *version 1* 
related to the inserted motif *RQWW* and 13 motifs related to motifs shared by
the clonally expanded sequences *CATSRAAKPDGLAALETQYF* and 
*CATSRAAKPDGLAALSTQYF*.

#### Version 2 vs. Version 1

The last comparison will be between *version 1* and *version 2*.

```{r local_motifs_v2_vs_v1, fig.height=7, fig.width=11}
# select all motifs that are passed in at least one version of v1 and v2
local_motifs_v2_v1 <- local_motifs[(local_motifs$pass_v1 == TRUE |
                                      local_motifs$pass_v2 == TRUE),]

# display with ggplot
ggplot(local_motifs_v2_v1, aes(-log10(p_value_v2),
                               -log10(p_value_v1),
                               label = motif,
                               color = pass_v2)) +
  geom_point(size=3) +
  geom_label_repel(nudge_x = 3, force = 100, max.overlaps = Inf, 
                   show.legend = FALSE, size = 7) +
  scale_colour_manual(name="Found motifs",
                      label=c("v1 only (6)", 
                              "both v1 & v2 (6)"),
                      values = c("red", 
                                 "black")) +
  labs(x="-log10(p-value v2)", y = "-log10(p-value v1)") +
  theme(text = element_text(size = 23)) +
  lims(x = c(0,50), y=c(0,30))
```

The inserted motif *RQWW* gets detected with a high p-value by both versions 
(marked black in the graph), as well as the related motifs *QWWA*, *QWWG*, 
*QWWS*, *QRQW* and *YRQW*. Motif *SYRQ* gets only found by *version 1*. This is
caused by underestimation of the number of found motifs in the reference 
(`mean_f_ref`) by *version 1*, which is calculated as the mean of repeated 
bootstrapping, leading in effect to a "happy accident" in which *SYRQ* is found 
by chance. Also *version 2* calculates a slightly higher false discovery rate 
value (*0.05268845*) than the cutoff of *0.05*. Reducing the sequences to a 
non-redundant set affects the number of motifs present in the data sample 
and reference sample, which is the reason why *SYRQ* is found by *version 3* 
but not by *version 2*.

No shared motifs between clonally expanded sequences are marked as enriched, 
because *version 1* and *version 2* both reduce non-redundant sequences before 
looking for motifs.

Five motifs are only found by *version 1* (marked as red in the graph).
The found 4mers *GFNT*, *GGSY*, *LVSG*, *SLVS*, and *SNRA* are neither related
to the artificial enrichment of motif *RQWW* nor to the two clonally expanded 
sequences *CATSRAAKPDGLAALETQYF* and *CATSRAAKPDGLAALSTQYF*. Therefore, these 
motifs represent false positives caused by the p-value precision of 
*version 1* being too low.

In total, *version 1* finds 12 motifs, of which five are false positives.
*Version 2* finds six motifs.  Thirteen motifs between the clonally expanded
sequences are neither detected by *version 1* nor *version 2*.

### Global clustering output - all versions

Inspecting the global motif clustering shows the same output for all versions.
The first pair of CDR3b sequences (*CASSYVLRAADGYTF* and *CASSYVLRAADGDTF*) 
were present in the reference dataset and were randomly sampled into the 
example dataset, while the second pair (*CATSRAAKPDGLAALSTQYF* and 
*CATSRAAKPDGLAALETQYF*) represents the artificially enhanced clonal expansion.

Again, if the sampling was repeated with a different seed, other random 
pairs of the reference dataset could be sampled. The two added sequences
are guaranteed to be found by all three algorithm versions.


```{r global_motifs_output}
# global motifs version 1
kable(clustirr_output_v1$clust$CDR3b$global)

# global motifs version 2
kable(clustirr_output_v2$clust$CDR3b$global)

# global motifs version 3
kable(clustirr_output_v3$clust$CDR3b$global)
```
### Graph output

TBA

As mentioned in the *user manual vignette*, graph output is planned to be added
in future `r Biocpkg("ClustIRR")` versions.

# References

**1**: Glanville, Jacob, et al. "Identifying specificity groups in the 
T cell receptor repertoire." Nature 547.7661 (2017): 94.<br>
**2**: Huang, Huang, et al. "Analyzing the Mycobacterium tuberculosis immune 
response by T-cell receptor clustering with GLIPH2 and genome-wide antigen 
screening." Nature Biotechnology 38.10 (2020): 1194-1202.<br>

# Session Info

```{r package_version}
packageVersion("ClustIRR")
```


```{r session_info}
sessionInfo()
```
